File: rust/tokio/src/bin/channel.rs
// Test performance of an async MPMC channel primitive.
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/channel.cpp
//
// N producers each post a slice of `element_count` items; M consumers pull
// until the channel is drained; the total count and sum are validated.
//
// tokio itself only ships an MPSC channel, so this benchmark uses `flume`, the
// runtime-agnostic async MPMC channel most tokio users reach for.

use std::time::Instant;
use tokio_bench::{multi_thread_runtime, peak_memory_usage};

const ELEMENT_COUNT: u64 = 10_000_000;
const ITER_COUNT: usize = 1;

struct Counts {
    count: u64,
    sum: u64,
}

// Split ELEMENT_COUNT into per-producer (count, base) work assignments.
fn producer_assignments(producer_count: usize) -> Vec<(u64, u64)> {
    let per = ELEMENT_COUNT / producer_count as u64;
    let rem = ELEMENT_COUNT % producer_count as u64;
    let mut base = 0u64;
    let mut out = Vec::with_capacity(producer_count);
    for i in 0..producer_count as u64 {
        let count = if i < rem { per + 1 } else { per };
        out.push((count, base));
        base += count;
    }
    out
}

async fn do_bench(producer_count: usize, consumer_count: usize) -> u64 {
    let (tx, rx) = flume::unbounded::<u64>();

    let mut producers = Vec::with_capacity(producer_count);
    for (count, base) in producer_assignments(producer_count) {
        let tx = tx.clone();
        producers.push(tokio::spawn(async move {
            for i in 0..count {
                tx.send_async(base + i).await.unwrap();
            }
        }));
    }
    drop(tx); // once all producer clones finish, consumers see the channel close

    let mut consumers = Vec::with_capacity(consumer_count);
    for _ in 0..consumer_count {
        let rx = rx.clone();
        consumers.push(tokio::spawn(async move {
            let mut count = 0u64;
            let mut sum = 0u64;
            while let Ok(v) = rx.recv_async().await {
                count += 1;
                sum += v;
            }
            Counts { count, sum }
        }));
    }
    drop(rx);

    for p in producers {
        p.await.unwrap();
    }
    let mut totals = Counts { count: 0, sum: 0 };
    for c in consumers {
        let r = c.await.unwrap();
        totals.count += r.count;
        totals.sum += r.sum;
    }

    let expected_sum = ELEMENT_COUNT * (ELEMENT_COUNT - 1) / 2;
    if totals.count != ELEMENT_COUNT {
        println!(
            "FAIL: Expected {} elements but consumed {} elements",
            ELEMENT_COUNT, totals.count
        );
    }
    if totals.sum != expected_sum {
        println!(
            "FAIL: Expected {} sum but got {} sum",
            expected_sum, totals.sum
        );
    }
    totals.sum
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let thread_count = if args.len() > 1 {
        args[1].parse().expect("thread count")
    } else {
        tokio_bench::default_threads()
    };

    let per = (thread_count / 2).max(1);
    let producer_count = per;
    let consumer_count = per;

    let rt = multi_thread_runtime(thread_count);
    println!("threads: {}", thread_count);
    println!("producers: {}", producer_count);
    println!("consumers: {}", consumer_count);

    rt.block_on(async move {
        let result = do_bench(producer_count, consumer_count).await; // warmup
        println!("output: {}", result);

        let start = Instant::now();
        for _ in 0..ITER_COUNT {
            let _ = do_bench(producer_count, consumer_count).await;
        }
        let dur_us = start.elapsed().as_micros();

        let elements_per_sec = (ELEMENT_COUNT as u128) * 1_000_000 / dur_us.max(1);
        println!("runs:");
        println!("  - iteration_count: {}", ITER_COUNT);
        println!("    elements: {}", ELEMENT_COUNT);
        println!("    duration: {} us", dur_us);
        println!("    elements/sec: {}", elements_per_sec);
        println!("    max_rss: {} KiB", peak_memory_usage());
    });
}

File: nim/weave/channel.nim
# Test performance of an MPMC channel primitive.
#
# Canonical (TooManyCooks) implementation:
# ../../cpp/TooManyCooks/channel.cpp
#
# Weave has no public MPMC channel (its internal channels are scheduler
# plumbing), and its workers must never block, so this benchmark measures
# Nim's official MPMC channel package - threading/channels
# (https://github.com/nim-lang/threading) - with one OS thread per producer
# and consumer. The harness labels the result "weave_threading" via
# benchmark_configs in build_and_bench_all.py, like tokio_flume.
#
# N producers each post a slice of elementCount items; M consumers pull until
# the channel is drained; the total count and sum are validated. Like the Go
# port, the channel is bounded (threading/channels has no unbounded mode) with
# a generous buffer so the benchmark measures steady-state throughput.
# threading/channels also has no close(): after the producers are joined, the
# main thread sends one sentinel per consumer to terminate them. The sentinel
# value is outside the produced range of [0, elementCount).

import std/[monotimes, times]
import threading/channels
import benchutil

const elementCount = 10_000_000
const iterCount = 1
const channelBuffer = 1 shl 16
const sentinel = high(uint64)

type Counts = object
  count: uint64
  sum: uint64

type ProducerArg = object
  chan: Chan[uint64]
  count: uint64
  base: uint64

type ConsumerArg = object
  chan: Chan[uint64]
  res: ptr Counts

proc producer(arg: ProducerArg) {.thread.} =
  for i in 0'u64 ..< arg.count:
    arg.chan.send(arg.base + i)

proc consumer(arg: ConsumerArg) {.thread.} =
  var count = 0'u64
  var sum = 0'u64
  var v: uint64
  while true:
    arg.chan.recv(v)
    if v == sentinel:
      break
    inc count
    sum += v
  arg.res[] = Counts(count: count, sum: sum)

proc doBench(producerCount, consumerCount: int): uint64 =
  var chan = newChan[uint64](channelBuffer)

  # Split elementCount into per-producer (count, base) assignments.
  var producers = newSeq[Thread[ProducerArg]](producerCount)
  let per = uint64(elementCount) div uint64(producerCount)
  let rem = uint64(elementCount) mod uint64(producerCount)
  var base = 0'u64
  for i in 0 ..< producerCount:
    var count = per
    if uint64(i) < rem:
      inc count
    createThread(producers[i], producer,
                 ProducerArg(chan: chan, count: count, base: base))
    base += count

  var results = newSeq[Counts](consumerCount)
  var consumers = newSeq[Thread[ConsumerArg]](consumerCount)
  for i in 0 ..< consumerCount:
    createThread(consumers[i], consumer,
                 ConsumerArg(chan: chan, res: addr results[i]))

  # Once every producer has finished, all real values precede the sentinels in
  # the FIFO, so each consumer drains its share and stops on its sentinel.
  joinThreads(producers)
  for _ in 0 ..< consumerCount:
    chan.send(sentinel)
  joinThreads(consumers)

  var total = Counts()
  for r in results:
    total.count += r.count
    total.sum += r.sum

  let expectedSum = uint64(elementCount) * (uint64(elementCount) - 1) div 2
  if total.count != uint64(elementCount):
    echo "FAIL: Expected ", elementCount, " elements but consumed ",
         total.count, " elements"
  if total.sum != expectedSum:
    echo "FAIL: Expected ", expectedSum, " sum but got ", total.sum, " sum"
  result = total.sum

proc main() =
  let threadCount = threadCountArg(1)

  var per = threadCount div 2
  if per < 1:
    per = 1
  let producerCount = per
  let consumerCount = per

  echo "threads: ", threadCount
  echo "producers: ", producerCount
  echo "consumers: ", consumerCount

  let res = doBench(producerCount, consumerCount) # warmup
  echo "output: ", res

  let start = getMonoTime()
  for _ in 0 ..< iterCount:
    discard doBench(producerCount, consumerCount)
  var durUs = (getMonoTime() - start).inMicroseconds
  if durUs < 1:
    durUs = 1
  let elementsPerSec = int64(elementCount) * 1_000_000 div durUs

  echo "runs:"
  echo "  - iteration_count: ", iterCount
  echo "    elements: ", elementCount
  echo "    duration: ", durUs, " us"
  echo "    elements/sec: ", elementsPerSec
  echo "    max_rss: ", peakMemoryUsageKiB(), " KiB"

when isMainModule:
  main()
